bpf-sys/src/type_gen.rs

```diff
@@ -0,0 +1,253 @@
+            return Err(TypeGenError::IO(io::Error::last_os_error()));
+        }
+        let header_name = outfile.as_ref().file_name().unwrap();
+        let guard_name = header_name
+            .to_str()
+            .unwrap()
+            .to_uppercase()
+            .chars()
+            .map(|c| match c {
+                'A'..='Z' => c,
+                _ => '_',
+            })
+            .collect::<String>();
+        unsafe {
+            let guard = CString::new(format!(
+                "#ifndef __{guard_name}__\n#define __{guard_name}__\n\n",
+                guard_name = guard_name
+            ))
+            .unwrap();
+            libc::fwrite(
+                guard.as_ptr() as *const _,
+                1,
+                guard.as_bytes().len(),
+                stream,
+            );
+        }
+        let stream = StreamWrapper(stream);
+        let white_re = if let Some(ref whitelist) = self.whitelist {
+            Some(RegexSet::new(whitelist).or(Err(TypeGenError::RegexError))?)
+        } else {
+            None
+        };
+        unsafe {
+            let dump_opts = {
+                let mut uninit = MaybeUninit::<btf_dump_opts>::zeroed();
+                (*uninit.as_mut_ptr()).ctx = stream.0 as *mut _;
+                uninit.assume_init()
+            };
+            let dumpptr = btf_dump__new(
+                self.btfptr,
+                ptr::null(),
+                &dump_opts as *const _,
+                Some(vfprintf_wrapper),
+            );
+            if (dumpptr as isize) < 0 {
+                return Err(TypeGenError::DumpError);
+            }
+            let dumpptr = BtfDumpWrapper(dumpptr);
+            for type_id in 1..=btf__get_nr_types(self.btfptr) {
+                let btftypeptr = btf__type_by_id(self.btfptr, type_id);
+                let nameptr = btf__name_by_offset(self.btfptr, (*btftypeptr).name_off);
+                if nameptr.is_null() {
+                    continue;
+                }
+
+                let cname = CStr::from_ptr(nameptr);
+                let namestr = cname.to_str().or(Err(TypeGenError::DumpError))?;
+                if let Some(wre) = &white_re {
+                    if !wre.is_match(namestr) {
+                        continue;
+                    }
+                }
+                if btf_dump__dump_type(dumpptr.0, type_id) < 0 {
+                    return Err(TypeGenError::DumpError);
+                }
+            }
+        }
+
+        unsafe {
+            let guard = CString::new("#endif\n").unwrap();
+            libc::fwrite(
+                guard.as_ptr() as *const _,
+                1,
+                guard.as_bytes().len(),
+                stream.0,
+            );
+        }
+
+        Ok(())
+    }
+}
+
+impl Drop for VmlinuxBtfDump {
+    fn drop(&mut self) {
+        unsafe {
+            btf__free(self.btfptr);
+        }
+    }
+}
+
+// wrapping vfprintf to get rid of return type
+unsafe extern "C" fn vfprintf_wrapper(
```

Comment:
I feel like this probably can go as well, if the code above uses `std::fs` instead of the `libc` api?